src/socket/mod.rs

```diff
@@ -0,0 +1,311 @@
+        W: serde::Serialize,
+    {
+        let (timeout, write_res) =
+            with_timeout(timeout, |t| self.write(t, mes)).map_err(|()| WriteError::Timeout)?;
+        write_res?;
+        let e = self.read(&timeout)?;
+        Ok(e)
+    }
+
+    /// Listen for a message from the other side and immediately
+    /// send a reply based on the message.
+    ///
+    /// The timeout counts for the whole roundtrip.
+    pub fn react<F>(&mut self, timeout: Timeout, reaction: F) -> Result<R, ReadWriteError>
+    where
+        R: serde::de::DeserializeOwned,
+        W: serde::Serialize,
+        F: FnOnce(&R) -> W,
+    {
+        let (timeout, read_res) =
+            with_timeout(timeout, |t| self.read(t)).map_err(|()| ReadError::Timeout)?;
+        let read = read_res?;
+        self.write(&timeout, &reaction(&read))?;
+        Ok(read)
+    }
+
+    /// Check if the underlying socket timed out when serializing/deserilizing.
+    fn is_timed_out(e: &bincode::ErrorKind) -> bool {
+        match e {
+            bincode::ErrorKind::Io(io) => match io.kind() {
+                std::io::ErrorKind::TimedOut => true,
+                _ => false,
+            },
+            _ => false,
+        }
+    }
+
+    /// Wait for a message to arrive.
+    pub fn read(&self, timeout: &Timeout) -> Result<R, ReadError>
+    where
+        R: serde::de::DeserializeOwned,
+    {
+        let timeout_socket = timeout::TimeoutReadWriter::new(self.socket, timeout);
+
+        // XXX: “If this returns an Error, `reader` may be in an invalid state”.
+        // what the heck does that mean.
+        bincode::deserialize_from(timeout_socket).map_err(|e| {
+            if Self::is_timed_out(&e) {
+                ReadError::Timeout
+            } else {
+                ReadError::Deserialize(e)
+            }
+        })
+    }
+
+    /// Send a message to the other side.
+    pub fn write(&mut self, timeout: &Timeout, mes: &W) -> Result<(), WriteError>
+    where
+        W: serde::Serialize,
+    {
+        let timeout_socket = timeout::TimeoutReadWriter::new(self.socket, timeout);
+
+        bincode::serialize_into(timeout_socket, mes).map_err(|e| {
+            if Self::is_timed_out(&e) {
+                WriteError::Timeout
+            } else {
+                WriteError::Serialize(e)
+            }
+        })?;
+
+        into_bincode_io_error(self.socket.flush())?;
+
+        Ok(())
+    }
+}
+
+/// Wrap a socket with a timeout. Inspired by https://docs.rs/crate/timeout-readwrite/0.2.0/.
+mod timeout {
+    extern crate nix;
+
+    use self::nix::libc;
+    use self::nix::poll;
+    use super::{Millis, Timeout};
+    use std::os::unix::io::AsRawFd;
+    use std::os::unix::net::UnixStream;
+
+    /// Wait until `to_fd` receives the poll event from `events`, up to `timeout` length
+    /// of time.
+    /// Copied from https://docs.rs/crate/timeout-readwrite/0.2.0/source/src/utils.rs
+    /// written by Jonathan Creekmore and published under Apache-2.0.
```

Comment:
mmm maybe we should use that crate?

Reply:
Nah, it has a quite horrible interface and we just need this one bit of logic, so I’d rather copy that with a link back.